server/athenian/api/controllers/miners/github/contributors.py

```diff
@@ -21,94 +23,109 @@ async def mine_contributors(repos: Collection[str],
+            .where(and_(PullRequestReview.repository_full_name.in_(repos),
+                        PullRequestReview.submitted_at.between(time_from, time_to)))
+            .group_by(PullRequestReview.user_login))
 
     @sentry_span
-    async def fetch_commit_authors():
+    async def fetch_commit_author():
         return await db.fetch_all(
             select([PushCommit.author_login, func.count(PushCommit.author_login)])
             .where(and_(PushCommit.repository_full_name.in_(repos),
                         PushCommit.committed_date.between(time_from, time_to)))
             .group_by(PushCommit.author_login))
 
     @sentry_span
-    async def fetch_commit_committers():
+    async def fetch_commit_committer():
         return await db.fetch_all(
             select([PushCommit.committer_login, func.count(PushCommit.committer_login)])
             .where(and_(PushCommit.repository_full_name.in_(repos),
                         PushCommit.committed_date.between(time_from, time_to)))
             .group_by(PushCommit.committer_login))
 
     @sentry_span
-    async def fetch_reviews():
+    async def fetch_commenter():
         return await db.fetch_all(
-            select([PullRequestReview.user_login, func.count(PullRequestReview.user_login)])
-            .where(and_(PullRequestReview.repository_full_name.in_(repos),
-                        PullRequestReview.submitted_at.between(time_from, time_to)))
-            .group_by(PullRequestReview.user_login))
+            select([PullRequestComment.user_login, func.count(PullRequestComment.user_login)])
+            .where(and_(PullRequestComment.repository_full_name.in_(repos),
+                        PullRequestComment.created_at.between(time_from, time_to),
+                        ))
+            .group_by(PullRequestComment.user_login))
 
     @sentry_span
-    async def fetch_releases():
+    async def fetch_merger():
+        return await db.fetch_all(
+            select([PullRequest.merged_by_login, func.count(PullRequest.merged_by_login)])
+            .where(common_prs_where)
+            .group_by(PullRequest.merged_by_login))
+
+    @sentry_span
+    async def fetch_releaser():
         return await db.fetch_all(
             select([Release.author, func.count(Release.author)])
             .where(and_(Release.repository_full_name.in_(repos),
                         Release.published_at.between(time_from, time_to)))
             .group_by(Release.author))
 
-    data = await asyncio.gather(
-        fetch_prs(), fetch_comments(), fetch_commit_authors(), fetch_commit_committers(),
-        fetch_reviews(), fetch_releases())
-
+    as_roles = as_roles or ("author", "reviewer", "commit_author", "commit_committer",
+                            "commenter", "merger", "releaser")
+    locals_ = locals()
+    fetchers = {k: locals_[f"fetch_{k}"]() for k in as_roles}
+    data = await asyncio.gather(*fetchers.values())
```

Comment:
We can do this with less hacks :smile: 
```suggestion
    @sentry_span
    async def fetch_author():
        return "prs", await db.fetch_all(...)

    tasks = [fetch_author(), ..., fetch_releaser()]
    data = await asyncio.gather(*tasks, return_exceptions=True)
    for r in data:
        if isinstance(r, Exception):
            raise r from None
    for key, rows in data:
        ...
```

Reply:
The idea here is to avoid calling specific `fetch_*` functions if they're not necessary, that is a `fetch_X` function is called if `X` is in `as_roles` or if `as_roles` is empty. Anyway, I can achieve the same without using `locals()` if this is what bothers you 👍, but I'd still need the mapping between the role name and the function and to filter the functions to call depending on the `as_roles` parameter.

I'll add the `return_exceptions` 👍

Reply:
I am against any `locals()` here. The code becomes hostile to refactorings, IDE intellisense, etc. Explicit is better than implicit. Besides, this is closer to the Go way.

Reply:
As you see, you don't need the mapping if you return it from each function.

Reply:
> I am against any locals() here. The code becomes hostile to refactorings, IDE intellisense, etc. Explicit is better than implicit. Besides, this is closer to the Go way.

👍 

> As you see, you don't need the mapping if you return it from each function.

I cannot make it return from each function because I need to filter them before calling them in order to avoid fetching unnecessary things.